### PR TITLE
Features/rest hive client

### PR DESCRIPTION
**Pull request by marcoooo**
First Hive Module to be used when needing a Python standard REST client within a Hive Pipeline.
Usage:
Inherit from the base class:

override `run()` method to process response data
override `write_output()` to pass results to following pipeline items.

### CODE REVIEW

**Comment by muffato** [comment]
Sorry I haven't had time to review this yet. I will address the changes that are required in a PR or directly push the change

**Comment by marcoooo** [comment]
> Sorry I haven't had time to review this yet. I will address the changes that are required in a PR or directly push the change

Do you mean there was a lot of changes "required" ?

File: .codecov.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+coverage:
+  status:
+    project:
+      default:
+        # Don't report failures on pull-requests
+        informational: true
```
Comment (muffato) [inline .codecov.yml:6]: Missing end-of-line character. Will push the fix directly to master

File: src/tests/hive/rest.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import requests_mock
+from eHive.Process import Job
+
+from ensembl.hive.HiveRESTClient import HiveRESTClient
+
+in_pipe = open('hive.in', mode='rb', buffering=0)
+out_pipe = open('hive.out', mode='wb', buffering=0)
+
+
+class TestHiveRest(object):
```
Comment (muffato) [inline src/tests/hive/rest.py:31]: I'm working on a Python-only Runnable tester. Will be ready soon

File: .codecov.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (marcoooo) [reply .codecov.yml:6]: Yep not "yet".

File: src/python/ensembl/hive/HiveRESTClient.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    """
+    available_method = ('post', 'get', 'put', 'patch')
+
+    def param_defaults(self):
+        """
+        Default parameter set
+        :return: dict
+        """
+        return {
+            'endpoint': 'http://localhost/api/endpoint',
```
[the remark this answers is not in the document]
Comment (muffato) [reply src/python/ensembl/hive/HiveRESTClient.py:40]: Done in #2

File: src/python/ensembl/hive/HiveRESTClient.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        return http
+
+    def _close_session(self, session):
+        """
+        Close all potential remaining connections in current Session
+        :return None
+        """
+        session.close()
+
+    @contextlib.contextmanager
```
[the remark this answers is not in the document]
Comment (muffato) [reply src/python/ensembl/hive/HiveRESTClient.py:74]: Done in #2

File: src/python/ensembl/hive/HiveRESTClient.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+                                         headers=self.param('headers'),
+                                         files=self.param('files'),
+                                         data=self.param('payload'),
+                                         timeout=self.param('timeout'))
+
+
+
+    def run(self):
+        # should do something with results from fetch_input method
+        self.warning("You may do something with retrieved response {}".format(self.response.json()), is_error=False)
```
[the remark this answers is not in the document]
Comment (muffato) [reply src/python/ensembl/hive/HiveRESTClient.py:106]: Done in #2

File: src/python/ensembl/hive/HiveRESTClient.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+            self._close_session(session)
+
+    def fetch_input(self):
+        """
+        Basic call to request parameters specified in pipeline parameters
+        Return response received.
+        """
+        with self._session_scope() as http:
+            self.response = http.request(method=self.param('method'),
+                                         url=self.param('endpoint'),
```
[the remark this answers is not in the document]
Comment (muffato) [reply src/python/ensembl/hive/HiveRESTClient.py:96]: Done in #2

File: src/python/ensembl/hive/HiveRESTClient.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+            logger.debug("Closing session")
+            self._close_session(session)
+
+    def fetch_input(self):
+        """
+        Basic call to request parameters specified in pipeline parameters
+        Return response received.
+        """
+        with self._session_scope() as http:
+            self.response = http.request(method=self.param('method'),
```
[the remark this answers is not in the document]
Comment (muffato) [reply src/python/ensembl/hive/HiveRESTClient.py:95]: Done in #2

File: src/tests/hive/rest.py (hunk 2 above)
Comment (muffato) [reply src/tests/hive/rest.py:31]: See #2

**Pull request merged**